core/shared/src/main/scala/fs2/async/Promise.scala

```diff
@@ -0,0 +1,167 @@
+package fs2.async
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+import cats.implicits.{ catsSyntaxEither => _, _ }
+import cats.effect.{ Sync, Effect, IO }
+
+import fs2.Scheduler
+import fs2.internal.LinkedMap
+import java.util.concurrent.atomic.{AtomicBoolean,AtomicReference}
+
+import Promise._
+
+// TODO shall we move Effect and EC implicits back up here?
+// TODO scaladoc
+final class Promise[F[_], A] private [fs2] (ref: Ref[F, State[A]]) {
+
+  def get(implicit F: Effect[F], ec: ExecutionContext): F[A] =
+    F.delay(new MsgId).flatMap(getOrWait)
+
+  // TODO I prefer to not have `setAsync`, and leave forking at call site,
+  // since we can't exploit SyncRef lazy set directly
+  // I've replaced any setAsyncPure in fs2 with fork(setSync), which would
+  // probably reveal that there's a bunch of unnecessary forks that can be avoided
+  //
+  // NOTE: this differs in behaviour from the old Ref in that by the time readers are notified
+  // the new value is already guaranteed to be in place.
+  def setSync(a: A)(implicit F: Effect[F], ec: ExecutionContext): F[Unit] = {
```

Comment:
I agree with not providing `setAsync`. I think we can rename this to `set` now too.